yarn-project/sequencer-client/src/sequencer/sequencer.ts

```diff
@@ -475,6 +479,69 @@ export class Sequencer {
     return true;
   }
 
+  /**
+   * Build a block
+   *
+   * Shared between the sequencer and the validator for re-execution
+   *
+   * @param validTxs - The valid transactions to construct the block from
+   * @param newGlobalVariables - The global variables for the new block
+   * @param historicalHeader - The historical header of the parent
+   * @param interrupt - The interrupt callback, used to validate the block for submission and check if we should propose the block
+   */
+  private async buildBlock(
+    validTxs: Tx[],
+    newGlobalVariables: GlobalVariables,
+    historicalHeader?: Header,
+    interrupt?: (processedTxs: ProcessedTx[]) => Promise<void>,
+  ) {
+    this.log.debug('Requesting L1 to L2 messages from contract');
+    const l1ToL2Messages = await this.l1ToL2MessageSource.getL1ToL2Messages(newGlobalVariables.blockNumber.toBigInt());
+    this.log.verbose(
+      `Retrieved ${l1ToL2Messages.length} L1 to L2 messages for block ${newGlobalVariables.blockNumber.toNumber()}`,
+    );
+
+    const numRealTxs = validTxs.length;
+    const blockSize = Math.max(2, numRealTxs);
+
+    // Sync to the previous block at least
+    await this.worldState.syncImmediate(newGlobalVariables.blockNumber.toNumber() - 1);
+
+    // NB: separating the dbs because both should update the state
+    const publicProcessorFork = await this.worldState.fork();
+    const orchestratorFork = await this.worldState.fork();
+
+    // We create a fresh processor each time to reset any cached state (eg storage writes)
+    const processor = this.publicProcessorFactory.create(publicProcessorFork, historicalHeader, newGlobalVariables);
+    const blockBuildingTimer = new Timer();
+    const blockBuilder = this.blockBuilderFactory.create(orchestratorFork);
+    await blockBuilder.startNewBlock(blockSize, newGlobalVariables, l1ToL2Messages);
+
+    const [publicProcessorDuration, [processedTxs, failedTxs]] = await elapsed(() =>
+      processor.process(
+        validTxs,
+        blockSize,
+        blockBuilder,
+        this.txValidatorFactory.validatorForProcessedTxs(publicProcessorFork),
+      ),
+    );
+    if (failedTxs.length > 0) {
+      const failedTxData = failedTxs.map(fail => fail.tx);
+      this.log.debug(`Dropping failed txs ${Tx.getHashes(failedTxData).join(', ')}`);
+      await this.p2pClient.deleteTxs(Tx.getHashes(failedTxData));
+    }
+
+    await interrupt?.(processedTxs);
+
+    // All real transactions have been added, set the block as full and complete the proving.
+    const block = await blockBuilder.setBlockCompleted();
+
+    await publicProcessorFork.close();
```

Comment:
I don't know how these forks work, but previously it was the case that `close` was occurring in a `finally`. Is this change safe?

Reply:
it should stay in a finally, if the interrupt ends up throwing an error the db connection would remain open, will adjust